subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessorTest.groovy

```diff
@@ -80,6 +81,63 @@ class ForkingTestClassProcessorTest extends Specification {
         1 * workerProcessBuilder.setImplementationClasspath(_) >> { assert it[0].size() == 16 }
     }
 
+    def "stopNow propagates to worker process"() {
+        TestClassRunInfo testClass = Mock()
+        def remoteProcessor = Mock(RemoteTestClassProcessor)
+
+        setup:
+        1 * workerProcessFactory.create(_) >> workerProcessBuilder
+        1 * workerProcessBuilder.build() >> workerProcess
+        _ * workerProcessBuilder.getJavaCommand() >> Stub (JavaExecHandleBuilder)
+        1 * workerProcess.getConnection() >> Stub(ObjectConnection) { addOutgoing(_) >> remoteProcessor }
+
+        when:
+        processor.processTestClass(testClass)
+        processor.stopNow()
+
+        then:
+        10 * moduleRegistry.getModule(_) >> { module(it[0]) }
+        6 * moduleRegistry.getExternalModule(_) >> { module(it[0]) }
+        1 * workerLease.startChild()
+        1 * options.getSystemProperties() >> [:]
+        1 * workerProcess.stopNow()
+    }
+
+    def "stopNow does nothing when no remote processor"() {
+        when:
+        processor.stopNow()
+
+        then:
+        1 * processor.stopNow()
+        0 * _
+    }
+
+    def "no exception when stop after stopNow"() {
+        TestClassRunInfo testClass = Mock()
+        def remoteProcessor = Mock(RemoteTestClassProcessor)
+        WorkerLeaseRegistry.WorkerLeaseCompletion completion = Mock()
+
+        setup:
+        1 * workerProcessFactory.create(_) >> workerProcessBuilder
+        1 * workerProcessBuilder.build() >> workerProcess
+        _ * workerProcessBuilder.getJavaCommand() >> Stub (JavaExecHandleBuilder)
+        1 * workerProcess.getConnection() >> Stub(ObjectConnection) { addOutgoing(_) >> remoteProcessor }
+
+        when:
+        processor.processTestClass(testClass)
+        processor.stopNow()
+        processor.stop()
+
+        then:
+        1 * workerLease.startChild() >> completion
```

Comment:
❌ This applies to some of the other tests as well: This is over-asserting. Only mock the interactions that you are actually trying to test, ideally one per test. Otherwise you'll bind the test so tightly to the implementation that it'll fail for all kinds of unrelated reasons, making refactoring almost impossible.

Reply:
I'm at a bit of a loss here.  Most of the mock interactions here are needed because they are returning objects(/mocks) that are used by the `ForkingTestClassProcessor`, either in managing its internal state or coordinating the underlying worker process.  Any concrete suggestions?

Reply:
Right now this test is asserting:

- how many modules a worker process loads
- how the worker gets its system properties
- that the worker will request a worker lease
- that the worker comes from a builder that comes from a factory

All of these are things that this test is not about. The majority of these interactions could be replaced by a `Stub` instead of a `Mock`. Don't let the current structure of the test class pin you down in that regard. E.g. you could replace most of fields at the top with `Stub`s and tighten the assertions in the tests where they really matter.

Reply:
Makes more sense now.  I've reworked the test methods (in particular the ones around `stopNow`) to use `Stub()`s more & far fewer assertions on `Mock()`s.